risc0/zkvm/src/host/server/exec/syscall/fork.rs

```diff
@@ -0,0 +1,333 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{bail, Context as _, Result};
+use risc0_circuit_rv32im::prove::emu::{
+    addr::{ByteAddr, WordAddr},
+    rv32im::{DecodedInstruction, EmuContext, Emulator, Instruction, TrapCause},
+};
+use risc0_zkvm_platform::{
+    memory::is_guest_memory,
+    syscall::{
+        ecall,
+        reg_abi::{REG_A0, REG_A1, REG_A2, REG_MAX, REG_T0},
+    },
+    PAGE_SIZE, WORD_SIZE,
+};
+
+use super::{Syscall, SyscallContext, SyscallTable};
+
+const PAGE_WORDS: usize = PAGE_SIZE / WORD_SIZE;
+const INVALID_IDX: u32 = u32::MAX;
+const NUM_PAGES: usize = 192 * 1024;
+
+const PID_CHILD: u32 = 0;
+const PID_PARENT: u32 = 1;
+
+pub(crate) struct SysFork;
+
+impl Syscall for SysFork {
+    fn syscall(
+        &mut self,
+        _syscall: &str,
+        ctx: &mut dyn SyscallContext,
+        _to_guest: &mut [u32],
+    ) -> Result<(u32, u32)> {
+        let mut exec = ChildExecutor::new(ctx)?;
+        exec.run()?;
+        // Return a non-zero value for sys_fork, which means this is the 'parent' process.
+        Ok((PID_PARENT, 0))
+    }
+}
+
+struct Page(Vec<u8>);
+
+struct ChildExecutor<'a, 'b> {
+    ctx: &'b mut dyn SyscallContext<'a>,
+    pc: ByteAddr,
+    registers: [u32; REG_MAX],
+    page_table: Vec<u32>,
+    page_cache: Vec<Page>,
+    halt: bool,
+    syscall_table: SyscallTable<'a>,
+}
+
+impl<'a, 'b> ChildExecutor<'a, 'b> {
+    pub fn new(ctx: &'b mut dyn SyscallContext<'a>) -> Result<Self> {
+        // Start the child process one instruction past the ecall.
+        let pc = ByteAddr(ctx.get_pc() + WORD_SIZE as u32);
+
+        let mut registers = std::array::from_fn(|idx| ctx.load_register(idx));
+        // The return value of sys_fork is the pid, so set a0 to PID_CHILD.
+        registers[REG_A0] = PID_CHILD;
+
+        let syscall_table = ctx.syscall_table().clone();
```

Comment:
This code is pretty new to me so I might be wrong here but it seems like this allows the child to inherit the parent process's syscall and I'm wondering if this means that it will allow the child to do things like write to the journal. Writing to the journal will result in a failure but I'm not sure when it will fail. It would be nice to catch something like this during the execution of the child process..

Reply:
> > if this means that it will allow the child to do things like write to the journal
> 
> Given that `fork(2)` makes child processes inherit all file descriptors, I believe this would be expected behavior

Great call out - I didn't really think of it this way

> The intention with inheriting the syscall table is to allow disabling of syscalls in the child, such as to prevent fork bombs. Currently syscall data is coupled with its logic. Perhaps we should make a clear separation for data that gets copied vs copy-on-write vs shared?

I think this is a great approach! - maybe it might be good to allow SYS_WRITE to stdout/stderr for easier debugging but do some copy on write or no-op for journal? Anyway, thanks for the response, this was very helpful

Reply:
I pushed an update that removes the JOURNAL fd from the child, it also removes the sys_fork entry to prevent fork bombs.